File: katas/php/SOLID/ocp/ocp-good.php
<?php
// Cumplimiento del OCP: Abierto para extensión, cerrado para modificación
// Solución: Usar interfaces para que se puedan agregar nuevos animales sin cambiar el código existente

interface Communicable {
    public function communicate(): string;
}

class Dog implements Communicable {
    public function communicate(): string {
        return "woof woof";
    }
}

class Cat implements Communicable {
    public function communicate(): string {
        return "meow meow";
    }
}

// ✅ Esta clase nunca necesita ser modificada al agregar nuevos animales
class Communication {
    public function communicate(Communicable $animal): string {
        return $animal->communicate();
    }

    public function communicateMultiple(array $animals): array {
        return array_map(fn($animal) => $animal->communicate(), $animals);
    }
}

// ✅ Agregar nuevos animales sin modificar el código existente
class Fox implements Communicable {
    public function communicate(): string {
        return "ring-ding-ding-ding-dingeringeding";
    }
}

class Cow implements Communicable {
    public function communicate(): string {
        return "moo moo";
    }
}

class Duck implements Communicable {
    public function communicate(): string {
        return "quack quack";
    }
}

// Uso - ¡agregar Fox, Cow, Duck NO requirió cambios en las clases existentes!
$communication = new Communication();
$dog = new Dog();
$cat = new Cat();
$fox = new Fox();
$cow = new Cow();
$duck = new Duck();

echo $communication->communicate($dog) . PHP_EOL; // "woof woof"
echo $communication->communicate($cat) . PHP_EOL; // "meow meow"
echo $communication->communicate($fox) . PHP_EOL; // "ring-ding-ding-ding-dingeringeding"
echo $communication->communicate($cow) . PHP_EOL; // "moo moo"
echo $communication->communicate($duck) . PHP_EOL; // "quack quack"

$allAnimals = [$dog, $cat, $fox, $cow, $duck];
print_r($communication->communicateMultiple($allAnimals));

File: katas/csharp/SOLID/ocp/ocp-good.cs
using System;
using System.Collections.Generic;

// Cumplimiento del OCP: Abierto para extensión, cerrado para modificación
// ✅ Solución: Usar interfaces para que se puedan agregar nuevos animales sin cambiar el código existente

public interface ICommunicable
{
  string Communicate();
}

public class Dog : ICommunicable
{
  public string Communicate()
  {
    return "woof woof";
  }
}

public class Cat : ICommunicable
{
  public string Communicate()
  {
    return "meow meow";
  }
}

// ✅ Esta clase nunca necesita ser modificada al agregar nuevos animales
public class Communication
{
  public string Communicate(ICommunicable animal)
  {
    return animal.Communicate();
  }

  public List<string> CommunicateMultiple(List<ICommunicable> animals)
  {
    var results = new List<string>();
    foreach (var animal in animals)
    {
      results.Add(animal.Communicate());
    }
    return results;
  }
}

// ✅ Agregar nuevos animales sin modificar el código existente
public class Fox : ICommunicable
{
  public string Communicate()
  {
    return "ring-ding-ding-ding-dingeringeding";
  }
}

public class Cow : ICommunicable
{
  public string Communicate()
  {
    return "moo moo";
  }
}

public class Duck : ICommunicable
{
  public string Communicate()
  {
    return "quack quack";
  }
}

// Uso - ¡agregar Fox, Cow, Duck NO requirió cambios en las clases existentes!
class Program
{
  static void Main()
  {
    var communication = new Communication();
    var dog = new Dog();
    var cat = new Cat();
    var fox = new Fox();
    var cow = new Cow();
    var duck = new Duck();

    Console.WriteLine(communication.Communicate(dog)); // "woof woof"
    Console.WriteLine(communication.Communicate(cat)); // "meow meow"
    Console.WriteLine(communication.Communicate(fox)); // "ring-ding-ding-ding-dingeringeding"
    Console.WriteLine(communication.Communicate(cow)); // "moo moo"
    Console.WriteLine(communication.Communicate(duck)); // "quack quack"

    var allAnimals = new List<ICommunicable> { dog, cat, fox, cow, duck };
    var results = communication.CommunicateMultiple(allAnimals);
    Console.WriteLine(string.Join(", ", results));
  }
}
